### ISSUE TITLE
 unused_io_amount false positive if the result is implicitly return in a if/else.

### ISSUE THREAD

**Issue by mgautierfr** [rust-lang/rust-clippy#12519]
### Summary

Clippy report a false positive for `unused_io_amount` when the amount is implicitly returned at end of if/else block.

### Lint Name

unused_io_amount

### Reproducer

I tried this code:

```rust
use std::io::*;

struct MyBuffer<T: Read> {
    src: T,
    current: usize,
}

impl<T: Read> Read for MyBuffer<T> {
    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
        let read_size = if self.current < 100 {
            self.src.read(buf)?
        } else {
            self.src.read(&mut buf[..100])?
        };
        self.current += read_size;
        Ok(read_size)
    }
}

fn main() -> Result<()> {
    let buffer = vec![];
    let mut buffer = MyBuffer {
        src: Cursor::new(buffer),
        current: 0,
    };

    copy(&mut buffer, &mut stdout())?;
    Ok(())
}
```

I saw this happen:

```
$ clippy-driver test_read.rs
error: read amount is not handled
  --> test_read.rs:11:13
   |
11 |             self.src.read(buf)?
   |             ^^^^^^^^^^^^^^^^^^^
   |
   = help: use `Read::read_exact` instead, or handle partial reads
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_io_amount
   = note: `#[deny(clippy::unused_io_amount)]` on by default

error: read amount is not handled
  --> test_read.rs:13:13
   |
13 |             self.src.read(&mut buf[..100])?
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: use `Read::read_exact` instead, or handle partial reads
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_io_amount

error: aborting due to 2 previous errors

```

I expected to see this happen:

No error as read amount is actually handled.

Using the following leads to a (expected) `let_and_return` warning :

```rust
use std::io::*;

struct MyBuffer<T: Read> {
    src: T,
    current: usize,
}

impl<T: Read> Read for MyBuffer<T> {
    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
        let read_size = if self.current < 100 {
            let r = self.src.read(buf)?;
            r
        } else {
            let r = self.src.read(&mut buf[..100])?;
            r
        };
        self.current += read_size;
        Ok(read_size)
    }
}

fn main() -> Result<()> {
    let buffer = vec![];
    let mut buffer = MyBuffer {
        src: Cursor::new(buffer),
        current: 0,
    };

    copy(&mut buffer, &mut stdout())?;
    Ok(())
}
```

### Version

```text
$ rustc -Vv
rustc 1.77.0 (aedd173a2 2024-03-17)
binary: rustc
commit-hash: aedd173a2c086e558c2b66d3743b344f977621a7
commit-date: 2024-03-17
host: x86_64-unknown-linux-gnu
release: 1.77.0
LLVM version: 17.0.6


$ clippy-driver -V
clippy 0.1.77 (aedd173 2024-03-17)
```

**Comment by m-rph** [comment]
This was fixed shortly after, but unfurtonately the PR wasn't tagged to make sure it was merged within this release.
